src/partition/repart.c

```diff
@@ -3280,12 +3282,67 @@ static int do_make_directories(Partition *p, const char *root) {
         return 0;
 }
 
-static int partition_populate(Partition *p, const char *node) {
+static int partition_populate_directory(Partition *p, char **ret_root, char **ret_tmp_root) {
+        _cleanup_(rm_rf_physical_and_freep) char *root = NULL;
+        int r;
+
+        assert(ret_root);
+        assert(ret_tmp_root);
+
+        /* When generating squashfs, we need the source tree to be available when we generate the squashfs
+         * filesystem. Because we might have multiple source trees, we build a temporary source tree
+         * beforehand where we merge all our inputs. We then use this merged source tree to create the
+         * squashfs filesystem. */
+
+        if (!streq(p->format, "squashfs")) {
+                *ret_root = NULL;
+                *ret_tmp_root = NULL;
+                return 0;
+        }
+
+        /* If we only have a single directory that's meant to become the root directory of the filesystem,
+         * we can shortcut this function and just use that directory as the root directory instead. If we
+         * allocate a temporary directory, it's stored in `ret_tmp_root` to indicate it should be removed.
+         * Otherwise, we return the directory to use in `root` to indicate it should not be removed. */
```

Comment:
Please don't use backticks for quotes. This isn't markdown, so it's not rendered in any special way, and backticks just look bad.